tests/integration/test_integration_artifactory_path.py

```diff
@@ -173,6 +175,81 @@ def test_deploy_file(path):
     p.unlink()
 
 
+def test_deploy_file_by_checksum(path):
+    p = path("/integration-artifactory-path-repo/foo")
+    p1 = path("/integration-artifactory-path-repo/bar")
+
+    if p.exists():
+        p.unlink()
+    if p1.exists():
+        p1.unlink()
+
+    tf = tempfile.NamedTemporaryFile()
+    tf.write(b"Some test string")
+    tf.flush()
+
+    sha1 = sha1sum(tf.name)
+    sha256 = sha256sum(tf.name)
+
+    sha1_non_existent = "1111111111111111111111111111111111111111"
+    sha256_non_existent = (
+        "1111111111111111111111111111111111111111111111111111111111111111"
+    )
+
+    p.deploy_file(tf.name)
+    tf.close()
+    with p.open() as fd:
+        result = fd.read()
+    assert result == b"Some test string"
+
+    # The matrix is a full list of all possible combinations
+    # 1st row is for 1st parameter of deploy_by_checksum
+    # 2nd row is for 2nd parameter and 3rd is for 3rd parameter
+    # matrix = [
+    #     [sha1, sha1_non_existent, None],
```

Comment:
Can we handle "all available variants for tests" aka matrix via pytest.mark.parameterize? 
The test is quite bit right now and have a lot of tests inside, I think

Reply:
Hi @allburov 
It's a trade-off here. I just pick up some of cases, could you please give me some advice?
I tried to get the combination via `itertools.product`

Besides this specific function, there is same issue in other functions.

`deploy_by_checksum` accepts 4 parameters. Actually, each one of them has many possible values like below.

For simplicity, ignore `parameters` in `deploy_by_checksum`
```
matrix = [
    [sha1, sha1_non_existent, sha1_with_invalid_lenght, sha1_contains_invalid_charactors, None],
    [sha256, sha256_non_existent, sha256_with_invalid_lenght, sha256_contains_invalid_charactors, None],
    [sha1, sha1_non_existent, sha256, sha256_non_existent, sha1_or_sha256_with_invalid_lenght, sha1_or_sha256_contains_invalid_charactors, None]
]
```

There would be 5 * 5 * 7 = 175 cases. I tried to test each cases, but it has not obvious regularity, that's to say, it's difficult for us to assert the actual result is expected. And no [official statement](https://www.jfrog.com/confluence/display/RTF6X/Artifactory+REST+API#ArtifactoryRESTAPI-DeployArtifactbyChecksum) to guide us.

Reply:
You can specify only these parameters in pytest.mark.paramatrize and not to run 175 cases. 

I'm talking about the test - it has too many repetitive parts, it should be one test and params

```python
@pytest.mark.parametrize(
    "deploy_by_checksum_kwargs, error",
    [
        pytest.param(dict(sha1="sha1here"), None, id="SHA1 - file found"),
        pytest.param(dict(sha256="sha256here"), None, id="SHA256 - file found"),
        pytest.param(dict(sha1="sha1-not-exist"), RuntimeError, id="SHA1 - file not found"),
    ],
)

def test_deploy_file_by_checksum(path, deploy_by_checksum_kwargs, error):
    ctx = pytest.raises(error) if error else NullContext()
   with ctx:
      p.deploy_by_checksum(**deploy_by_checksum_kwargs)
   if not error:
      # asserts here
```

Reply:
Hi @allburov 

Thanks for your advice.

If we introduce `pytest.mark.parametrize`, it does avoid the repetitive parts.

But we introduce `if` statement, developers need to know where they are and what is the real context when error occurs.

Let me know what you think

Reply:
@zhan9san 
I think with pytest you get explicit error saying what was the parameter set that failed: https://docs.pytest.org/en/6.2.x/parametrize.html#pytest-mark-parametrize-parametrizing-test-functions

however, I would not use `deploy_by_checksum_kwargs` as proposed, since this will not be really easy to understand what we are calling, I would use explicit names of arguments of the function
but just use 3 decorators for each of the parameter, like in pytest example
~~~python
@pytest.mark.parametrize("x", [0, 1])
@pytest.mark.parametrize("y", [2, 3])
def test_foo(x, y):
    pass
~~~

Reply:
we need to finish testing part and I want to merge this PR to develop before we merge develop to master

Reply:
Hi @beliaev-maksim  and @allburov 

The original issue finding in this code review is 'too many repetitive parts', if I understand correctly.

Compared with unit test, there are same issues, with which I think it's acceptable.

[aql.create_aql_text](https://github.com/devopshq/artifactory/blob/develop/tests/unit/test_artifactory_path.py#L1120-L1148)
[check](https://github.com/devopshq/artifactory/blob/develop/tests/unit/test_artifactory_path.py#L80-L182)

IMHO, the duplicate code make it easier to identify which line acts as not expected.

Reply:
@zhan9san 
on this I agree with you

but you mentioned that we have 175 possible variation, of course, we cannot "hard code" them.
So parametric run will cover more scenarios compared to ~8 scenarios that present

Reply:
As I mentioned before, 

> I tried to test each cases, but it has not obvious regularity, that's to say, it's difficult for us to assert the actual result is expected.

it is required to introduce 'if' statements.

Do we really want to do this?

Besides pytest.mark.parametrize, I tried to use 'itertools.product' which can generate all possible values as well

Reply:
unfortunately, I do not have local sandbox artifactory to run integration tests. 

I would expect that artifactory provides same response for all cases when 
1. sha length is wrong
2. sha has wrong chars
3. sha valid

they potentially, we can build a map of input and expected output

if you have artifactory sandbox, then you can check this behaviour. There might be a need to sync parameters (not any to any, but go column by column in a matrix)

Reply:
As you can see in current test assertions, it's not as simple as 
AssertEqual when get expected success or AssertNotEqual when get expected failure.

The for-loop does avoid the duplicate code when the input and output are easy to be unified. In this specific case, the verification of output is not easy to be unified. technically, it can be.

For testing code itself, the more logic we introduce, the harder for us to debug.

Due to my lack of knowledge, correct me if I say anything wrong.